examples/webpack-react-macros/src/Usecases/ElementAttributes.js

```diff
@@ -1,38 +1,38 @@
 // @flow
 import * as React from "react"
-import { withI18n } from "@lingui/react"
-import { Trans } from "@lingui/react.macro"
-import { t } from "@lingui/js.macro"
+import { I18n } from "@lingui/react"
+import { t, Trans } from "@lingui/macro"
 
 type ElementAttributesProps = {
   i18n: Object
 }
 
-class ElementAttributes extends React.Component<ElementAttributesProps> {
+export default class ElementAttributes extends React.Component<
+  ElementAttributesProps
+> {
   props: ElementAttributesProps
 
   render() {
-    const { i18n } = this.props
     const articleName = "Scientific Journal"
     const closeLabel = t`Close`
 
     return (
-      <div>
-        <Trans>
-          <a
-            className="expression"
-            href="/article"
-            title={t`Full content of ${articleName}`}
-          >
-            Article
-          </a>
-        </Trans>
-        <button className="variable" aria-label={closeLabel}>
-          X
-        </button>
-      </div>
+      <I18n>
+        {({ i18n }) => (
+          <div>
+            <a
+              className="expression"
+              href="/article"
+              title={i18n._(t`Full content of ${articleName}`)}
```

Comment:
Is it really an improvement considering how it was looking before? :) Was the previous approach working? I am not sure, haven't used that. Also, I think that in gitter we were talking about a variant like `i18n(t`Full content of ${articleName}`)` which would much nicer imo.

Reply:
You mean `i18n.t` over `i18n._(t`? It's necessary to move to macros... I don't think there's a way to have a object property to be a macro.

Other way would be convert it to `i18n._` directly, but for that we would need `i18n` in scope, which wouldn't be used otherwise and linter would complain.

This approach is very flexible, but the only drawback is how it looks :/ But you really get used to it. I wrote `vanilla-js` example and it was so great.

`i18n(t` is planned, I just have to figure out how to create callable object. I already started working on it but then figure out this is not the core of the problem and can be done later. Both `i18n._(..)` and `i18n(...)` will be supported anyway until v3.

Reply:
And yeah, the previous approach had several problems, especially when plugins were misconfigured. It doesn't allow you to use `this.props.i18n.t`, because only `i18n.t` worked. You couldn't use it with `i18n.use` method, which switches language locally (`i18n.use('cs')._(t...)`). And also for *lazy messages* you had to use different API, `i18nMark`. Now there's the same API for all messages, lazy/immediate translations, with/without IDs.

As I said, I like everything about it except how it looks... but I believe it won't be such a problem once we have `i18n(t...)` API.

Reply:
Well, a "callable object" is just a function with additional properties/methods attached directly to it. There is nothing wrong about that in my opinion and it will make it much easier to use and look at :)

Reply:
Cool! I believe there's nothing wrong with it as well, but it felt a bit hackish :) I'm halfway through with implementation, which I saved in different branch. I'll finalize it soon.